components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,193 @@
+      /*.num_errors_to_ignore = */ 0,
+      /*.initial_delay_ms = */
+      features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.multiply_factor =*/2.0,
+      /*.jitter_factor =*/0.0,
+      /*.maximum_backoff_ms =*/16 *
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.always_use_initial_delay =*/true);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  StopParticipating();
+  eligibility_service_->RemoveObserver(this);
+}
+
+void LearningService::Init() {
+  DCHECK(url_loader_factory_);
+  DCHECK(eligibility_service_);
+  DCHECK(init_task_timer_);
+  communication_adapter_ = new CommunicationAdapter(url_loader_factory_);
+  eligibility_service_->AddObserver(this);
+
+  StartParticipating();
+}
+
+void LearningService::StartParticipating() {
+  participating_ = true;
+
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  DCHECK(participating_);
+  participating_ = false;
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, base::Unretained(this)));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  // net::BackoffEntry handle_or_reconnect_backoff_entry =
+  //   net::BackoffEntry(&handle_or_reconnect_policy_);
+  // handle_or_reconnect_backoff_entry.InformOfRequest(tasks.size());
+
+  if (tasks.size() == 0) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "No tasks available, reconnecting in " << reconnect << "s";
+    return;
+  }
+
+  // TODO(lminto): for now, disregard all tasks beyond the first one
+  // TODO(stevelaskaridis): extract into a utility function
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  ModelSpec spec{
+      32,   // num_params
+      64,   // batch_size
+      lr,   // learning_rate
+      500,  // num_iterations
+      0.5   // threshold
+  };
+
+  if (spec.num_params != static_cast<int>(task.GetParameters().at(0).size())) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  std::unique_ptr<Model> model = std::make_unique<Model>(spec);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
```

Comment:
could `GetParameters()` have length <= 1? Or `GetParameters().at(1)` be empty?

Reply:
in `flower_helper.cc` function at line 29 `ParseTaskListFromResponseBody`, we added some input validation. And if the parameters field are empty we discard the whole task